File: drake_ros_systems/example/rs_flip_flop.cpp
#include <drake/systems/analysis/simulator.h>
#include <drake/systems/framework/diagram_builder.h>
#include <drake/systems/framework/leaf_system.h>

#include <drake_ros_systems/drake_ros.hpp>
#include <drake_ros_systems/ros_interface_system.hpp>
#include <drake_ros_systems/ros_publisher_system.hpp>
#include <drake_ros_systems/ros_subscriber_system.hpp>

#include <std_msgs/msg/bool.hpp>

using drake_ros_systems::DrakeRos;
using drake_ros_systems::RosInterfaceSystem;
using drake_ros_systems::RosPublisherSystem;
using drake_ros_systems::RosSubscriberSystem;

class NorGate : public drake::systems::LeafSystem<double>
{
public:
  NorGate()
  {
    DeclareAbstractInputPort("A", *drake::AbstractValue::Make(std_msgs::msg::Bool()));
    DeclareAbstractInputPort("B", *drake::AbstractValue::Make(std_msgs::msg::Bool()));

    DeclareAbstractOutputPort("Q", &NorGate::calc_output_value);
  }

  virtual ~NorGate() = default;

private:
  void
  calc_output_value(const drake::systems::Context<double> & context, std_msgs::msg::Bool * output) const
  {
    const bool a = GetInputPort("A").Eval<std_msgs::msg::Bool>(context).data;
    const bool b = GetInputPort("B").Eval<std_msgs::msg::Bool>(context).data;
    output->data = !(a || b);
  }
};

// Delay's input port by one timestep to avoid algebraic loop error
// Inspired by Simulink's Memory block
template <typename T>
class Memory : public drake::systems::LeafSystem<double>
{
public:
  Memory(const T & initial_value)
  {
    DeclareAbstractInputPort("value", *drake::AbstractValue::Make(T()));

    // State for value
    DeclareAbstractState(drake::AbstractValue::Make(initial_value));

    // Output depends only on the previous state
    DeclareAbstractOutputPort("value", &Memory::calc_output_value, {all_state_ticket()});

    DeclarePerStepEvent(
      drake::systems::UnrestrictedUpdateEvent<double>([this](
          const drake::systems::Context<double>& context,
          const drake::systems::UnrestrictedUpdateEvent<double>&,
          drake::systems::State<double> * state) {
        // Copy input value to state
        drake::systems::AbstractValues & abstract_state = state->get_mutable_abstract_state();
        abstract_state.get_mutable_value(0).SetFrom(
          get_input_port().Eval<drake::AbstractValue>(context));
      }));
  }

  virtual ~Memory() = default;

private:
  void
  calc_output_value(const drake::systems::Context<double> & context, T * output) const
  {
    *output = context.get_abstract_state().get_value(0).get_value<T>();
  }
};

int main()
{
  // NOR gate RS flip flop example
  // Input topics /S and /R are active high (true is logic 1 and false is logic 0)
  // Output topics  /Q and /Q_not are active low (true is logic 0 and false is logic 1)

  // Input/Output table
  // S: false R: false | Q: no change  Q_not: no change
  // S: true  R: false | Q: false      Q_not: true
  // S: false R: true  | Q: true       Q_not: false
  // S: true  R: true  | Q: invalid    Q_not: invalid
  drake::systems::DiagramBuilder<double> builder;

  rclcpp::QoS qos{10};

  auto sys_ros_interface = builder.AddSystem<RosInterfaceSystem>(std::make_unique<DrakeRos>());
  auto sys_pub_Q = builder.AddSystem(
    RosPublisherSystem::Make<std_msgs::msg::Bool>(
      "Q", qos, sys_ros_interface->get_ros_interface()));
  auto sys_pub_Q_not = builder.AddSystem(
    RosPublisherSystem::Make<std_msgs::msg::Bool>(
      "Q_not", qos, sys_ros_interface->get_ros_interface()));
  auto sys_sub_S = builder.AddSystem(
    RosSubscriberSystem::Make<std_msgs::msg::Bool>(
      "S", qos, sys_ros_interface->get_ros_interface()));
  auto sys_sub_R = builder.AddSystem(
    RosSubscriberSystem::Make<std_msgs::msg::Bool>(
      "R", qos, sys_ros_interface->get_ros_interface()));
  auto sys_nor_gate_1 = builder.AddSystem<NorGate>();
  auto sys_nor_gate_2 = builder.AddSystem<NorGate>();
  auto sys_memory = builder.AddSystem<Memory<std_msgs::msg::Bool>>(std_msgs::msg::Bool());

  builder.Connect(sys_nor_gate_1->GetOutputPort("Q"), sys_memory->get_input_port(0));

  builder.Connect(sys_sub_S->get_output_port(0), sys_nor_gate_1->GetInputPort("A"));
  builder.Connect(sys_nor_gate_2->GetOutputPort("Q"), sys_nor_gate_1->GetInputPort("B"));

  builder.Connect(sys_memory->get_output_port(0), sys_nor_gate_2->GetInputPort("A"));
  builder.Connect(sys_sub_R->get_output_port(0), sys_nor_gate_2->GetInputPort("B"));

  builder.Connect(sys_nor_gate_1->GetOutputPort("Q"), sys_pub_Q->get_input_port(0));
  builder.Connect(sys_nor_gate_2->GetOutputPort("Q"), sys_pub_Q_not->get_input_port(0));

  auto diagram = builder.Build();
  auto context = diagram->CreateDefaultContext();

  auto simulator = std::make_unique<drake::systems::Simulator<double>>(*diagram, std::move(context));
  simulator->set_target_realtime_rate(1.0);
  simulator->Initialize();

  auto & simulator_context = simulator->get_mutable_context();

  while (true) {
    simulator->AdvanceTo(simulator_context.get_time() + 0.1);
  }
  return 0;
}

File: drake_ros_systems/example/rs_flip_flop.py
#!/usr/bin/env python3

from drake_ros_systems import RosInterfaceSystem
from drake_ros_systems import RosPublisherSystem
from drake_ros_systems import RosSubscriberSystem

from pydrake.systems.analysis import Simulator
from pydrake.systems.framework import DiagramBuilder
from pydrake.systems.framework import UnrestrictedUpdateEvent
from pydrake.common.value import AbstractValue
from pydrake.systems.framework import LeafSystem

from std_msgs.msg import String
from std_msgs.msg import Bool


class NorGate(LeafSystem):

    def __init__(self):
        super().__init__()
        self._a = self.DeclareAbstractInputPort("A", AbstractValue.Make(Bool()))
        self._b = self.DeclareAbstractInputPort("B", AbstractValue.Make(Bool()))

        self.DeclareAbstractOutputPort(
            'Q',
            lambda: AbstractValue.Make(Bool()),
            self._calc_output_value)

    def _calc_output_value(self, context, data):
        a = self._a.Eval(context)
        b = self._b.Eval(context)
        data.get_mutable_value().data = not (a.data or b.data)


class Memory(LeafSystem):
    """Delay input port by one time step to avoid algebraic loop error."""

    def __init__(self, initial_value):
        super().__init__()

        self._input = self.DeclareAbstractInputPort("A", AbstractValue.Make(initial_value))

        self.DeclareAbstractState(AbstractValue.Make(initial_value))

        self.DeclareAbstractOutputPort(
            'Q',
            lambda: AbstractValue.Make(initial_value),
            self._calc_output_value,
            {self.all_state_ticket()})

        self.DeclarePerStepEvent(UnrestrictedUpdateEvent(self._move_input_to_state))

    def _move_input_to_state(self, context, event, state):
        state.get_mutable_abstract_state().get_mutable_value(0).SetFrom(
            AbstractValue.Make(self._input.Eval(context)))

    def _calc_output_value(self, context, output):
        output.SetFrom(context.get_abstract_state().get_value(0))


def main():
    # NOR gate RS flip flop example
    # Input topics /S and /R are active high (true is logic 1 and false is logic 0)
    # Output topics  /Q and /Q_not are active low (true is logic 0 and false is logic 1)

    # Input/Output table
    # S: false R: false | Q: no change  Q_not: no change
    # S: true  R: false | Q: false      Q_not: true
    # S: false R: true  | Q: true       Q_not: false
    # S: true  R: true  | Q: invalid    Q_not: invalid
    builder = DiagramBuilder()

    sys_ros_interface = builder.AddSystem(RosInterfaceSystem())

    sys_pub_Q = builder.AddSystem(
        RosPublisherSystem(Bool, "Q", sys_ros_interface.get_ros_interface()))
    sys_pub_Q_not = builder.AddSystem(
        RosPublisherSystem(Bool, "Q_not", sys_ros_interface.get_ros_interface()))

    sys_sub_S = builder.AddSystem(
        RosSubscriberSystem(Bool, "S", sys_ros_interface.get_ros_interface()))
    sys_sub_R = builder.AddSystem(
        RosSubscriberSystem(Bool, "R", sys_ros_interface.get_ros_interface()))

    sys_nor_gate_1 = builder.AddSystem(NorGate())
    sys_nor_gate_2 = builder.AddSystem(NorGate())

    sys_memory = builder.AddSystem(Memory(Bool()))

    builder.Connect(
        sys_nor_gate_1.GetOutputPort('Q'),
        sys_memory.get_input_port(0)
    )

    builder.Connect(
        sys_sub_S.get_output_port(0),
        sys_nor_gate_1.GetInputPort('A'),
    )
    builder.Connect(
        sys_nor_gate_2.GetOutputPort('Q'),
        sys_nor_gate_1.GetInputPort('B'),
    )

    builder.Connect(
        sys_memory.get_output_port(0),
        sys_nor_gate_2.GetInputPort('A'),
    )
    builder.Connect(
        sys_sub_R.get_output_port(0),
        sys_nor_gate_2.GetInputPort('B'),
    )

    builder.Connect(
        sys_nor_gate_1.GetOutputPort('Q'),
        sys_pub_Q.get_input_port(0)
    )
    builder.Connect(
        sys_nor_gate_2.GetOutputPort('Q'),
        sys_pub_Q_not.get_input_port(0)
    )

    diagram = builder.Build()
    simulator = Simulator(diagram)
    simulator_context = simulator.get_mutable_context()
    simulator.set_target_realtime_rate(1.0)

    while True:
        simulator.AdvanceTo(simulator_context.get_time() + 0.1)


if __name__ == '__main__':
    main()
